File: Kalorie/Kalorie/Core/Utils/LoadingState.swift
//
//  LoadingState.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation

enum LoadingState<T> {
    case idle
    case loading
    case loaded(T)
    case error(Error?)
}

extension LoadingState where T == Void {
    static let loaded: LoadingState<Void> = .loaded(())
}

extension LoadingState {
    var isLoading: Bool {
        if case .loading = self { return true }
        return false
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/utils/LoadingState.kt
package antoni.kalorie.core.utils

sealed interface LoadingState<out T> {
    data object Idle : LoadingState<Nothing>
    data object Loading : LoadingState<Nothing>
    data class Loaded<T>(val value: T) : LoadingState<T>
    data class Error(val error: Throwable?) : LoadingState<Nothing>

    companion object {
        val loaded: LoadingState<Unit> = Loaded(Unit)
    }
}

val LoadingState<*>.isLoading: Boolean
    get() = this is LoadingState.Loading
